keps/sig-node/4622-topologymanager-max-allowable-numa-nodes/README.md

```diff
@@ -0,0 +1,693 @@
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+### Rollout, Upgrade and Rollback Planning
+
+When feature a is not enabled or configured, its value is the default value. and the feature is fully contained in the kubelet, has no dependencies and rollback and upgrades both will affect only newly created pods.
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Rollout or rollout fail do not impact already running workloads, only impact the new workloads.
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+We have a metric which records the topology manager admission time: `kubelet_topology_manager_admission_time`.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+Rollout or upgrade do not impact already running workloads. We plan to add an e2e test for this in the furture.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+No.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+We add a metric: `kubelet_topology_manager_admission_time` for kubelet, which can be used to check if the setting is causing unacceptable performance drops.
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+Examine the kubelet configuration of a node to verify the existence of the feature gate and the utilization of the new policy option.
```

Comment:
Per the comment for this question, is it possible to use a metric here?  Could we add one?  If that's not possible, can we be as prescriptive as possible in explaining exactly how to check a workload via the API?  I.e an exact kubectl command for a method of using kube_state_metrics or something?

I always try to imagine a cluster admin needing do this in an emergency.. providing an exact command makes it super fast..

Reply:
In this specific case we can add a gauge metric reporting the values of configured max NUMA nodes. I literally don't know if this is a good practice, but this approach seems quick and straightforward to implement.

Reply:
I don't think a metric is appropriate. It is not a feature that is used by workloads. It does not affect running workloads in any way. This feature is a property of the kubelet, and only affects how long it takes to admit a pod into the system. As such, the appropriate place to discover if this is enabled or not is by looking at the kubelet config.

Reply:
This command could be used to pull the options and inspect them:
```
kubectl get --raw "/api/v1/nodes/<nodename>/proxy/configz" | jq '.kubeletconfig.TopologyManagerPolicyOptions'
```

Reply:
thanks, updated.